Add support for custom objective functions in UniversalFallback.

martinbiel opened the pull request:
Quick prototype that resolves #1085 

Tests are missing.

blegat commented on src/Utilities/universalfallback.jl:
```diff
@@ -54,11 +57,12 @@ function Base.show(io::IO, U::UniversalFallback)
 
 function MOI.is_empty(uf::UniversalFallback)
-    return MOI.is_empty(uf.model) && isempty(uf.constraints) &&
+    return MOI.is_empty(uf.model) && isnothing(uf.objective) && isempty(uf.constraints) &&
```
Don't use `isnothing` as it is not available in Julia v1.0

blegat commented on src/Utilities/universalfallback.jl:
```diff
@@ -2,17 +2,18 @@
-`model` to create the model `UniversalFallback(model)` supporting *any*
-constaint and attribute. This allows to have a specialized implementation in
-`model` for performance critical constraints and attributes while still
+`model` to create the model `UniversalFallback(model)` supporting *any* objective,
```
Technically, the objective is an attribute. The objective attribute needs a special handling by `UniversalFallback` but it's still an attribute.

blegat commented on src/Utilities/universalfallback.jl:
```diff
@@ -156,6 +160,9 @@ function MOI.delete(uf::UniversalFallback, vi::VI)
     for d in values(uf.varattr)
         delete!(d, vi)
     end
+    if !(uf.objective === nothing)
```
You can use `!==`

blegat commented on src/Utilities/universalfallback.jl:
```diff
@@ -268,6 +281,52 @@ function MOI.get(uf::UniversalFallback, listattr::MOI.ListOfConstraintAttributes
 end
 
+# Objective
+MOI.supports(uf::UniversalFallback, ::MOI.ObjectiveSense) = true
```
This is already covered by
https://github.com/JuliaOpt/MathOptInterface.jl/blob/74d74433c783a980fa118ffbda1fbe9f5f0f1d8c/src/Utilities/universalfallback.jl#L339

blegat commented on src/Utilities/universalfallback.jl:
```diff
@@ -268,6 +281,52 @@ function MOI.get(uf::UniversalFallback, listattr::MOI.ListOfConstraintAttributes
+        return uf.objective
+    end
+end
+MOI.supports(uf::UniversalFallback, ::MOI.ObjectiveFunction{F}) where F <: MOI.AbstractScalarFunction = true
```
This is already covered by
https://github.com/JuliaOpt/MathOptInterface.jl/blob/74d74433c783a980fa118ffbda1fbe9f5f0f1d8c/src/Utilities/universalfallback.jl#L339

blegat commented on src/Utilities/universalfallback.jl:
```diff
@@ -268,6 +281,52 @@ function MOI.get(uf::UniversalFallback, listattr::MOI.ListOfConstraintAttributes
+        uf.objective = nothing
+    else
+        uf.objective = copy(func)
+        uf.model.objectiveset = false
```
`uf.model` is not necessarily an `MOIU.AbstractModel`. Here, you can do:
```julia
sense = MOI.get(uf.model, MOI.ObjectiveSense())
MOI.set(uf.model, MOI.ObjectiveSense(), MOI.FEASIBILITY_SENSE) # clear objective function if any was set
MOI.set(uf.model, MOI.ObjectiveSense(), sense)
```

blegat commented:
Implementation looks good. Just missing tests

martinbiel commented:
I added some simple tests to `universalfallback.jl`, but more importantly I believe 
```
function MOI.supports(
    ::ModelForUniversalFallback{T},
    ::Type{MOI.ObjectiveFunction{MOI.ScalarAffineFunction{T}}}) where T
    return false
end
```
should give a lot of coverage from the unit tests. I also rebased the branch on master.
